base/bucket_gocb.go

```diff
@@ -1053,29 +1092,96 @@ func (bucket CouchbaseBucketGoCB) GetWithXattr(k string, xattrKey string, rv int
+
+	// If the doc exists, delete both the doc body and the xattrs in one single op
+	_, mutateErr := bucket.Bucket.MutateInEx(k, deleteFlags, gocb.Cas(0), uint32(0)).
+		RemoveEx(xattrKey, gocb.SubdocFlagXattr). // Remove the xattr
+		RemoveEx("", gocb.SubdocFlagNone).        // Delete the document body
+		Execute()
+
+	// If no error, or it was just a ErrSubDocSuccessDeleted error, we're done.
+	// ErrSubDocSuccessDeleted is a "success error" that means "operation was on a tombstoned document"
+	if mutateErr == nil || mutateErr == gocbcore.ErrSubDocSuccessDeleted {
+		LogTo("CRUD+", "No error or ErrSubDocSuccessDeleted.  We're done.")
+		return nil
+	}
+
+	LogTo("CRUD+", "Initial attempt to delete doc failed with error: %v.  Trying again, this time only deleting doc body.", mutateErr)
+
+	// Invoke the callback which has the ability to change the document state
+	if callback != nil {
+		callback(
+			bucket,
+			k,
+			xattrKey,
+		)
+	}
+
+	//  Do get w/ xattr in order to get cas
+	var retrievedVal map[string]interface{}
+	var retrievedXattr map[string]interface{}
+	getCas, err := bucket.GetWithXattr(k, xattrKey, &retrievedVal, &retrievedXattr)
+	if err != nil {
 		return err
 	}
 
-	_, deleteXattrErr := bucket.Bucket.MutateInEx(k, gocb.SubdocDocFlagAccessDeleted, removeCas, 0).
-		RemoveEx(xattrKey, gocb.SubdocFlagXattr).
+	// If the doc body is non-empty at this point, then give up because it seems that a doc update has been
+	// interleaved with the purge.  Return error to the caller and cancel the purge.
+	if len(retrievedVal) != 0 {
+		return fmt.Errorf("DeleteWithXattr was unable to delete the doc.  It might be the case that another update " +
+			"was received which resurrected the doc by adding a new revision, in which case this delete operation is " +
+			"considred as cancelled.")
+	}
+
+	// Cas-safe delete of just the XATTR
+	_, mutateErrDeleteXattr := bucket.Bucket.MutateInEx(k, deleteFlags, gocb.Cas(getCas), uint32(0)).
+		RemoveEx(xattrKey, gocb.SubdocFlagXattr). // Remove the xattr
 		Execute()
 
-	if deleteXattrErr != nil && deleteXattrErr != gocbcore.ErrSubDocSuccessDeleted {
-		return deleteXattrErr
+	// If no error, or it was just a ErrSubDocSuccessDeleted error, we're done.
+	// ErrSubDocSuccessDeleted is a "success error" that means "operation was on a tombstoned document"
+	if mutateErrDeleteXattr == nil || mutateErrDeleteXattr == gocbcore.ErrSubDocSuccessDeleted {
+		return nil
 	}
 
-	return nil
+	// If the cas-safe delete of XATTR fails, return an error to the caller.
+	// This might happen if there was a concurrent update interleaved with the purge (someone resurrected doc)
+	return fmt.Errorf("DeleteWithXattr was unable to delete the doc.  It might be the case that another update "+
+		"was received which resurrected the doc by adding a new revision, in which case this delete operation is "+
+		"considred as cancelled.  Underlying gocb bucket operation error: %v", mutateErrDeleteXattr)
+
```

Comment:
Same as above, on the 'might be the case' language.  I think we should stick to the specifics of what we know.